[MRG] Use joblib.Parallel for CalibratedClassifierCV.fit

jjerphan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #17097.

#### What does this implement/fix? Explain your changes.

`CalibratedClassifierCV` fit estimator sequentially.

This PR introduces `joblib.Parallel` to calibrate estimators in parallel in `CalibratedClassifierCV.fit`
 
Parameters have been added to the constructor of `CalibratedClassifierCV` to match the ones from `joblib.Parallel`.

#### Any other comments?

This is my first contribution to scikit-learn.

I took some time to read [the guidelines in the documentation](https://scikit-learn.org/stable/developers/contributing.html#contributing-code), but I haven't found the kinds of tests such a contribution needs.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jjerphan committed:
Use joblib.Parallel for CalibratedClassifierCV.fit

jjerphan committed:
Modify for PEP8 compliance

jnothman commented:
>  I haven't found the kinds of tests such a contribution needs.

Usually a combination of: provide a benchmark in discussion; and add a test checking that the result with parallelism is equal to the result without.

jjerphan committed:
Test parallel execution of CalibratedClassifierCV

jjerphan committed:
WIP: Benchmarks for CalibratedClassifierCV

jjerphan committed:
Adapt Benchmark for CalibratedClassifierCVBenchmark

jjerphan committed:
Change number of samples for benchmark

jjerphan commented:
I've proposed some benchmarks; I'll post the results once I got them executed..

Let me know if you need other ones.

jjerphan commented:
Results got on a dataset of 10 000 samples.

```
BenchmarkNJobsAlgo {'estimator_name': 'LogisticReg.', 'n_jobs': 1} 10 trials: 0.2859 ± 0.1318 s
BenchmarkNJobsAlgo {'estimator_name': 'LogisticReg.', 'n_jobs': 2} 10 trials: 0.2500 ± 0.3721 s
BenchmarkNJobsAlgo {'estimator_name': 'LogisticReg.', 'n_jobs': 4} 10 trials: 0.1737 ± 0.3121 s
BenchmarkNJobsAlgo {'estimator_name': 'LogisticReg.', 'n_jobs': 8} 10 trials: 0.1792 ± 0.1365 s
BenchmarkNJobsAlgo {'estimator_name': 'CART', 'n_jobs': 1} 10 trials: 0.4957 ± 0.0051 s
BenchmarkNJobsAlgo {'estimator_name': 'CART', 'n_jobs': 2} 10 trials: 0.4095 ± 0.2507 s
BenchmarkNJobsAlgo {'estimator_name': 'CART', 'n_jobs': 4} 10 trials: 0.3547 ± 0.2749 s
BenchmarkNJobsAlgo {'estimator_name': 'CART', 'n_jobs': 8} 10 trials: 0.3460 ± 0.1434 s
BenchmarkNJobsAlgo {'estimator_name': 'ExtraTrees', 'n_jobs': 1} 10 trials: 4.0427 ± 0.2084 s
BenchmarkNJobsAlgo {'estimator_name': 'ExtraTrees', 'n_jobs': 2} 10 trials: 3.0686 ± 0.2907 s
BenchmarkNJobsAlgo {'estimator_name': 'ExtraTrees', 'n_jobs': 4} 10 trials: 2.9841 ± 0.3987 s
BenchmarkNJobsAlgo {'estimator_name': 'ExtraTrees', 'n_jobs': 8} 10 trials: 2.4930 ± 0.2843 s
BenchmarkNJobsAlgo {'estimator_name': 'RandomForest', 'n_jobs': 1} 10 trials: 10.7586 ± 0.4182 s
BenchmarkNJobsAlgo {'estimator_name': 'RandomForest', 'n_jobs': 2} 10 trials: 7.1585 ± 0.2358 s
BenchmarkNJobsAlgo {'estimator_name': 'RandomForest', 'n_jobs': 4} 10 trials: 6.4878 ± 0.8545 s
BenchmarkNJobsAlgo {'estimator_name': 'RandomForest', 'n_jobs': 8} 10 trials: 5.0233 ± 0.2879 s


BenchmarkCV {'cv': 2} 10 trials: 0.0634 ± 0.0140 s
BenchmarkCV {'cv': 3} 10 trials: 0.1066 ± 0.0168 s
BenchmarkCV {'cv': 4} 10 trials: 0.1521 ± 0.0490 s
BenchmarkCV {'cv': 5} 10 trials: 0.1958 ± 0.0435 s


BenchmarkNJobsCV {'cv': 2, 'n_jobs': 1} 10 trials: 0.0586 ± 0.0044 s
BenchmarkNJobsCV {'cv': 2, 'n_jobs': 2} 10 trials: 0.1773 ± 0.3157 s
BenchmarkNJobsCV {'cv': 2, 'n_jobs': 4} 10 trials: 0.2049 ± 0.3752 s
BenchmarkNJobsCV {'cv': 2, 'n_jobs': 8} 10 trials: 0.1568 ± 0.1029 s
BenchmarkNJobsCV {'cv': 3, 'n_jobs': 1} 10 trials: 0.0886 ± 0.0096 s
BenchmarkNJobsCV {'cv': 3, 'n_jobs': 2} 10 trials: 0.1720 ± 0.2619 s
BenchmarkNJobsCV {'cv': 3, 'n_jobs': 4} 10 trials: 0.1681 ± 0.3032 s
BenchmarkNJobsCV {'cv': 3, 'n_jobs': 8} 10 trials: 0.1613 ± 0.1106 s
BenchmarkNJobsCV {'cv': 4, 'n_jobs': 1} 10 trials: 0.1140 ± 0.0070 s
BenchmarkNJobsCV {'cv': 4, 'n_jobs': 2} 10 trials: 0.1745 ± 0.2710 s
BenchmarkNJobsCV {'cv': 4, 'n_jobs': 4} 10 trials: 0.1608 ± 0.3285 s
BenchmarkNJobsCV {'cv': 4, 'n_jobs': 8} 10 trials: 0.1737 ± 0.1293 s
BenchmarkNJobsCV {'cv': 5, 'n_jobs': 1} 10 trials: 0.1511 ± 0.0177 s
BenchmarkNJobsCV {'cv': 5, 'n_jobs': 2} 10 trials: 0.1954 ± 0.2658 s
BenchmarkNJobsCV {'cv': 5, 'n_jobs': 4} 10 trials: 0.1687 ± 0.3171 s
BenchmarkNJobsCV {'cv': 5, 'n_jobs': 8} 10 trials: 0.1872 ± 0.1150 s
```

jjerphan committed:
Add plot to benchmarks

jjerphan committed:
Format with black

jjerphan committed:
Change barplot for boxplot

jjerphan commented:
Here are some results I got on my latest runs (cc @jnothman).

--- 

### Medium dataset benchmark (i.e. a dataset of shape `(10000, 10)`).

Results have been obtains on 20 sequential runs.
The default algorithm is `LogisticRegression`.

### Benchmark environment 

- Python: 3.8.2
- OS: Fedora 32
- CPU: i7-550U @ 2.4GHz (4VCores)
- RAM: 8GB

### Single-threaded algorithm
![1](https://user-images.githubusercontent.com/13029839/81468396-a1697900-91df-11ea-9a08-9cb183cb61cb.png)

### Multi-threaded algorithm

![2](https://user-images.githubusercontent.com/13029839/81468400-a62e2d00-91df-11ea-8252-291598ce2eed.png)

### Number of folds (algorithm used: 
![3](https://user-images.githubusercontent.com/13029839/81468401-a75f5a00-91df-11ea-97a4-d949b61a207a.png)

### Number of folds and number of jobs
![5](https://user-images.githubusercontent.com/13029839/81468404-a8908700-91df-11ea-83f4-0180c71bb944.png)

---

### Discussions

Paralleling the Calibration generally results in shorter running times for medium datasets.

This is not related to this contribution but this is still good information: increasing the number of folds can results in a longer running times. The joint influence of changing the number of folds and the number of jobs has to be studied.

jjerphan committed:
Outputs dataset shape

jjerphan committed:
Move and comment calibration function

The calibration function has to be defined outside of
CalibratedClassifierCV.fit for pickling reasons.

This defines it at the file level.

It has been renamed and documented accordingly.

jjerphan commented:
### Large dataset benchmark (i.e. a dataset of shape `(100000, 10)`).

Results have been obtains on 20 sequential runs.
The default algorithm is `LogisticRegression`.

### Benchmark environment 

- Python: 3.8.2
- OS: Fedora 32
- CPU: i7-5500U @ 2.4GHz (4VCores)
- RAM: 8GB

### Single-threaded algorithm
![1](https://user-images.githubusercontent.com/13029839/81470802-17281180-91ed-11ea-9c56-9c15ed379267.png)


### Multi-threaded algorithm

Too long to perform on my machine.

### Number of folds :
![2](https://user-images.githubusercontent.com/13029839/81470805-1d1df280-91ed-11ea-8d08-40ddf536c645.png)

### Number of folds and number of jobs
![3](https://user-images.githubusercontent.com/13029839/81470806-1e4f1f80-91ed-11ea-8c8e-03316a5f7701.png)

---

### Discussions

Paralleling the Calibration generally also results in shorter running times for larger datasets.

jjerphan committed:
PEP8: Fix E501 (line too long)

thomasjpfan commented:
@jjerphan Those look like good results. May you provide the code you use for benchmarking?

jjerphan commented:
@thomasjpfan: Thanks ! The code is in [`benchmarks/bench_calibration.py`](https://github.com/scikit-learn/scikit-learn/blob/42f9acb105c2e8da396d98ffadb964b338a56cc0/benchmarks/bench_calibration.py)

jjerphan commented:
@thomasjpfan: have you have time to perform those benchmark yourself?
I'll try to run them with on other configurations. My hypothesis is that making it parallel might help in some situations but could come with slow downs if bigger datasets are used.

jjerphan commented:
Here are some new benchmarks I've performed for singly-threaded algorithm.
![benchmarknjobssinglethreadalgo_1000_10_20](https://user-images.githubusercontent.com/13029839/83322528-be322300-a258-11ea-8ceb-598b062968b8.png)
![benchmarknjobssinglethreadalgo_1000_100_20](https://user-images.githubusercontent.com/13029839/83322531-c1c5aa00-a258-11ea-88de-c788c4bd460e.png)
![benchmarknjobssinglethreadalgo_10000_10_20](https://user-images.githubusercontent.com/13029839/83322532-c2f6d700-a258-11ea-8033-ebecaf5d0500.png)
![benchmarknjobssinglethreadalgo_10000_100_20](https://user-images.githubusercontent.com/13029839/83322534-c4280400-a258-11ea-8ff1-2b86fdba6539.png)
![benchmarknjobssinglethreadalgo_100000_10_20](https://user-images.githubusercontent.com/13029839/83322537-c5593100-a258-11ea-8fcd-f1972126cdc4.png)
![benchmarknjobssinglethreadalgo_100000_100_20](https://user-images.githubusercontent.com/13029839/83322538-c68a5e00-a258-11ea-84b0-c2ade01d3ceb.png)

For multi-threaded algorithms, as they are using `joblib.Parallel` there might be some `Parallel` context nesting hence results aren't necessarily better.

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -30,6 +31,52 @@
+    calibrated_classifier : estimator object
+        The calibrated estimator
+    """
+    if sample_weight is not None \
+            and supports_sw:
```
```suggestion
    if sample_weight is not and supports_sw:
```

jjerphan replied:
Did you wanted to mean this instead ?
```python
if sample_weight and supports_sw:
```

jjerphan replied:
Hum, this is an incorrect syntax, what did you want to indicated instead ?
I think that we still need the first version.

jjerphan replied:
d2eb739  got your suggestion reverted.

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -30,6 +31,52 @@
+                          classes,
+                          sample_weight=None):
+    """Calibrate estimator for a given dataset split.
+    Parameters
+    ----------
```
I would say we do not need to document the parameters in this case.

thomasjpfan commented on sklearn/tests/test_calibration.py:
```diff
@@ -124,6 +124,31 @@ def test_sample_weight():
 
+def test_parallel_execution():
+    """Test parallel calibration"""
+    n_samples = 100
+    X, y = make_classification(n_samples=2 * n_samples, n_features=6,
+                               random_state=42)
+    X_train, y_train = X[:n_samples], y[:n_samples]
+    X_test = X[n_samples:]
```
We can use:

```py
X, y = make_classification(random_state=42)
X_train, y_train, X_test, y_test = train_test_split(X, y, random_state=42)
``

thomasjpfan reviewed (commented):
Thank you for the additional benchmark results. I am wondering if we need the benchmark script in the PR itself. Usually including the benchmark in a gist or the discussion is enough.

> For multi-threaded algorithms, as they are using joblib.Parallel there might be some Parallel context nesting hence results aren't necessarily better.

That makes sense. It's nice that joblib handles most of the over subscription as well.

jjerphan committed:
Update sklearn/calibration.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Remove parameters from doc-string for _calibrate_classifier

jjerphan committed:
Adapt test to use train_test_split

jjerphan committed:
Revert "Update sklearn/calibration.py"

This reverts commit da4ff2543ad1484f0ecee61bc561500ff1b29091.

jjerphan committed:
Remove benchmark

It can be found online here as a gist:
 - https://gist.github.com/jjerphan/4d10b9aad5573245bbd93da2e3fb1f23

jjerphan committed:
Remove unused local variable

jjerphan commented:
I've removed the benchmark from the code base and published them as a gist here:
 - https://gist.github.com/jjerphan/4d10b9aad5573245bbd93da2e3fb1f23

jjerphan commented:
My current thoughts about those modifications are that we might prefer only run `CalibratedClassifierCV.fit` if the underlying classifier algorithm is single-threaded solely – if the algorithm is multi-thread and/or does rely on `joblib.Parallel`, we might run in some unusual context nesting which might makes the overall unusable.

If we take this path, I guess that we just need to find a proper way to infer if an algorithm is multi-threaded or not. As of now, I do not know how to properly get this information in this context.

thomasjpfan commented:
Since the default is serial, I think this PR is acceptable.

In the future, we can add an `'auto'` option once we have a way to determine what best `n_jobs` to use.

jjerphan commented:
OK, it just needs another reviewer I guess then. 🙂

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -75,6 +104,32 @@ class CalibratedClassifierCV(BaseEstimator, ClassifierMixin,
+        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
+        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
+        for more details.
+
```
And the rest of the new parameters.

```suggestion

        .. versionadded:: 0.24
        
```

jjerphan replied:
I guess this change should come just before `n_jobs` in the doc-string, right?

jjerphan replied:
Ah nevermind, I just have understood the syntax.

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -30,6 +31,34 @@
+    calibrated_classifier : estimator object
+        The calibrated estimator
+    """
+    if sample_weight is not None \
+            and supports_sw:
```
```suggestion
    if sample_weight is not None and supports_sw:
```

thomasjpfan commented on sklearn/tests/test_calibration.py:
```diff
@@ -124,6 +124,28 @@ def test_sample_weight():
+        cal_clf_sequential.fit(X_train, y_train)
+        probs_sequential = cal_clf_sequential.predict_proba(X_test)
+
+        assert np.equal(probs__parallel, probs_sequential).all()
```
```suggestion
        assert_allclose(probs_parallel, probs_sequential)
```

where the import is from: `from numpy.testing import assert_allclose`.

thomasjpfan commented on sklearn/tests/test_calibration.py:
```diff
@@ -124,6 +124,28 @@ def test_sample_weight():
+    for method in ['sigmoid', 'isotonic']:
+        base_estimator = LinearSVC(random_state=42)
+
+        cal_clf__parallel = CalibratedClassifierCV(base_estimator,
+                                                   method=method)
+        cal_clf__parallel.fit(X_train, y_train)
+        probs__parallel = cal_clf__parallel.predict_proba(X_test)
```
```suggestion
        cal_clf_parallel = CalibratedClassifierCV(base_estimator,
                                                   method=method, n_jobs=2)
        cal_clf_parallel.fit(X_train, y_train)
        probs_parallel = cal_clf_parallel.predict_proba(X_test)
```

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -30,6 +31,34 @@
 from .utils.validation import _deprecate_positional_args
 
 
+def _calibrate_classifier(estimator, X, y, train, test,
+                          supports_sw,
+                          method,
+                          classes,
+                          sample_weight=None):
```
Nit:

```suggestion
def _calibrate_classifier(estimator, X, y, train, test, supports_sw,
                          method, classes, sample_weight=None):
```

thomasjpfan reviewed (commented):
Please add an entry to the change log at `doc/whats_new/v0.24.rst` with tag ||. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

jjerphan committed:
Improve code style

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Improve code style

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Fix naming

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Fix style for PEP8 compliance

jjerphan committed:
Change array tests to use allclose

jjerphan committed:
Add ::versionadded on doc-string for 0.24

jjerphan committed:
Merge branch 'master' into calibratedclassifierCV_joblib_parallel

jjerphan committed:
Add entry for #17107

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -209,31 +271,31 @@ def fit(self, X, y, sample_weight=None):
+            parallel = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,
+                                pre_dispatch=self.pre_dispatch)
+
+            with parallel:
```
I do not think we need this context manager, since there is only a single call to `parallel` here. The context manager is used when we want to [reuse a pool of workers](https://joblib.readthedocs.io/en/latest/parallel.html#reusing-a-pool-of-workers)

jjerphan committed:
Use correct reStructuredText format

jjerphan committed:
Remove context manager for joblib.Parallel

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -30,6 +31,30 @@
+                                                  method=method,
+                                                  classes=classes)
+    sw = None if sample_weight is None else sample_weight[test]
+    calibrated_classifier.fit(X[test], y[test],
+                              sample_weight=sw)
```
```suggestion
    calibrated_classifier.fit(X[test], y[test], sample_weight=sw)
```

thomasjpfan reviewed (approved):
LGTM

jjerphan committed:
Improve code style

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan commented:
Hi @thomasjpfan, shall I ask someone to be a second reviewer for this PR? 🙂
Thank you!

jjerphan committed:
Merge branch 'master' into calibratedclassifierCV_joblib_parallel

sahinbatmaz commented:
Is it ready to be merged ? 🙂

jjerphan commented:
It is ready and just needs another approval in order to be merged.

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -32,6 +33,29 @@
 from .utils.validation import _deprecate_positional_args
 
 
+def _calibrate_classifier(estimator, X, y, train, test, supports_sw,
```
```suggestion
def _fit_calibrated_classifer(estimator, X, y, train, test, supports_sw,
```

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -32,6 +33,29 @@
 
+def _calibrate_classifier(estimator, X, y, train, test, supports_sw,
+                          method, classes, sample_weight=None):
+    """Calibrate estimator for a given dataset split.
```
```suggestion
    """Fit calibrated classifier for a given dataset split.
```

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -227,31 +288,30 @@ def fit(self, X, y, sample_weight=None):
+                                pre_dispatch=self.pre_dispatch)
+
+            self.calibrated_classifiers_ = parallel(delayed(
+                _calibrate_classifier)(clone(base_estimator),
```
```suggestion
                _fit_calibrated_classifer)(clone(base_estimator),
```

thomasjpfan reviewed (commented):
Minor name change.

thomasjpfan commented:
@glemaitre If you are interested, this should be a pair-straight forward review.

jjerphan committed:
Better wording for calibrated classifier fitting [1/3]

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Better wording for calibrated classifier fitting [2/3]

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Better wording for calibrated classifier fitting [3/3]

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
lint: Fix E128

jjerphan commented:
This message comes as a small follow-up: can this PR be merger? :)

cc @glemaitre.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -90,6 +114,38 @@ class CalibratedClassifierCV(ClassifierMixin,
             ``cv`` default value if None changed from 3-fold to 5-fold.
 
+    n_jobs : int, default=None
+        Number of jobs to run in parallel.
```
Please explain here how parallelism happens for this meta estimator. For instance:

"""
Base estimator clones are fitted in parallel across cross-validation iterations.
Therefore parallelism happens only when cv != "prefit".
"""

ogrisel reviewed (approved):
Overall this LGTM. Here are two suggestions for improvements:

jjerphan committed:
Parametrize method for testing the parallel execution of CalibratedClassifierCV

jjerphan committed:
Precise CalibratedClassifierCV parallelism

jjerphan committed:
fix: E501

jjerphan commented on sklearn/tests/test_calibration.py:
```diff
@@ -152,6 +153,28 @@ def test_sample_weight():
+    X, y = make_classification(random_state=42)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
+
+    for method in ['sigmoid', 'isotonic']:
```
Done.
There are other tests which are structured liked this one for the same class but that aren't affected by this PR. I get that they should be changed as well but maybe in a separate PR.

ogrisel replied:
Yes, better keep this PR focused to the main topic.

ogrisel commented:
Thank you very much for your contribution @jjerphan!

jjerphan commented:
Thank you for your time and reviews @jnothman and @ogrisel!
